WordPress/openverse-api
# Improve the thumbnail service to support compression and WEBP

## Fixes


Fixes #531 by @zackkrida 

## Description


This PR replaces [willnorris/imageproxy](https://github.com/willnorris/imageproxy) with [h2non/imaginary](https://github.com/h2non/imaginary), a faster, more feature-rich and more highly-rated library for image manipulations over HTTP. This library allows us to compress thumbnails and reduce bytes.

Now, we can control the dimensions of the thumbnail and the amount of compression using env vars (while also allowing the URL query params `full_size` and `compressed` override the defaults if needed).

To spark joy, this PR adds API documentation for the thumbnail endpoints (both audio and video) with all query params documented.

## Testing Instructions

0. Check out the PR and fetch thumbnails for some audio/image files. They should load normally.
1. Play with the `full_size` and `compressed` query params.
  - With `full_size` set to `'true'`, the image should be larger.
  - With `full_size` set to `'false'`, the image will be smaller.
  - Compression is by default set to the opposite of `full_size`, but can be overridden with the `compressed` param.

## Stats

Here are some randomly sampled cases.

| Full-size no-compression | Full-size compression | Thumbnail no-compression | Thumbnail compression |
|-|-|-|-|
| 307 kB | 84.1 kB | 223 kB | 30.6 kB |
| 385 kB | 96.1 kB | 279 kB | 39.9 kB |
| 387 kB | 108 kB | 278 kB | 42.9 kB |

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
